### ISSUE TITLE
Nan training/validation loss when using fp16

### ISSUE THREAD

**Issue by njaremko** [fastai/fastai#1551]
**Describe the bug**
When doing tabular learning in fp16 mode, `train_loss` and `valid_loss` in the plotted table are `nan`. This same code without `to_fp16()` works fine and displays loss numbers.

**Provide your installation details**
```text
=== Software === 
python        : 3.7.2
fastai        : 1.0.43.dev0
fastprogress  : 0.1.18
torch         : 1.0.0
nvidia driver : 415.27
torch cuda    : 10.0.130 / is available
torch cudnn   : 7401 / is enabled

=== Hardware === 
nvidia gpus   : 1
torch devices : 1
  - gpu0      : 7949MB | GeForce RTX 2080

=== Environment === 
platform      : Linux-4.20.5-arch1-1-ARCH-x86_64-with-arch
distro        : Arch Linux  
conda env     : Unknown
python        : /usr/bin/python
sys.path      : /home/njaremko/Development
/usr/lib/python37.zip
/usr/lib/python3.7
/usr/lib/python3.7/lib-dynload

/home/njaremko/.local/lib/python3.7/site-packages
/usr/lib/python3.7/site-packages
/home/njaremko/Development/scikit-image
/home/njaremko/.local/lib/python3.7/site-packages/IPython/extensions
/home/njaremko/.ipython
```
**To Reproduce**
```python3
from fastai import *
from fastai.tabular import *

path = '.'

dep_var = 'target'
cat_names = ['stuff']
cont_names = ['cont_stuff']
procs = [FillMissing, Categorify, Normalize]

df = pd.read_csv('train.csv')
data = (TabularList.from_df(df, path=path, cat_names=cat_names, cont_names=cont_names, procs=procs)
                           .random_split_by_pct()
                           .label_from_df(cols=dep_var, label_cls=FloatList)
                           .add_test(test)
                           .databunch())

learn = tabular_learner(data, layers=[200,100], metrics=rmse)
learn.to_fp16()
learn.fit_one_cycle(3)
```

**Expected behavior**
The learning table should display correct training and validation losses

**Screenshots**
[fp32 Screenshot](https://www.dropbox.com/s/vjwf9ichaj2rb2i/Screenshot_20190131_170418.png?raw=1)
[fp16 Screenshot](https://www.dropbox.com/s/81e64imsqq9mt2i/Screenshot_20190131_170758.png?raw=1)

**Comment by gcp** [comment]
Does changing the loss_scale factor in to_fp16() change anything?

**Comment by njaremko** [comment]
I tried halving and doubling the `loss_scale`, but still got nan

**Comment by gcp** [comment]
I was thinking of cutting it by a much larger factor, back to 1 or so. You can move back up if it fixes the problem.

Apex/amp is a NVIDIA plugin for pytorch that has dynamic loss scaling, it might be worth a try and seeing if it works instead of fastai's to_fp16. You can then print what factor it selected.

**Comment by njaremko** [comment]
Setting it to 1 (or using amp) fixes the issue. Maybe there should be warning printed when calculated loss is nan in fp16 mode?

**Comment by sgugger** [comment]
I'll implement the dynamic loss scaling sometime soon, it's in my TODO. In the meantime, a warning would be fine, yes.
